[Chong Kok Leong] iP

SethCKL opened the pull request:
### Cleese

> and how may i help you sir? 😃 

~~not boujee enough~~ want to own a _real butler_? 
This is the next **best** thing for you I guess It's

- text-based
- easy peasy lemon squeezy
- FREE

All you gotta do it

1. download [this](https://github.com/SethCKL/ip/releases/tag/A-Jar)
2. double click to run it
3. talk to you new butler
4. let him manage your life

Features:

- [x] Managing tasks
- [ ] GUI (Coming Soon)

Sample Code:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

SethCKL committed:
Level-1. Greet, Echo, Exit

SethCKL committed:
Level 2. Add, List

SethCKL committed:
Level 3. Mark as Done

SethCKL committed:
Level 4. ToDos, Events, Deadlines

SethCKL committed:
Level 6. Delete

SethCKL committed:
Level 5. Handle Errors

SethCKL committed:
Merge remote-tracking branch 'origin/master'

SethCKL committed:
Level-7: Save

SethCKL committed:
Level-8: Dates and Times

SethCKL committed:
Level-8: Dates and Times

SethCKL committed:
A-MoreOOP: Use More OOP

SethCKL committed:
A-MoreOOP: Use More OOP

SethCKL committed:
Branch Level-8

SethCKL committed:
Revert "A-MoreOOP: Use More OOP"

This reverts commit 25e570386e4a5c76b06a6ab2b14bf05d40ea24be.

SethCKL committed:
Merge remote-tracking branch 'origin/master'

SethCKL committed:
A-Packages: Organise into Packages

SethCKL committed:
Added Junit tests

SethCKL committed:
Merge branch 'branch-Level-8'

SethCKL committed:
no message

SethCKL committed:
Added Level-9

LapisRaider commented on src/main/java/Cleese/Cleese.java:
```diff
@@ -0,0 +1,43 @@
+            ui.printLine();
+            } catch(Exception error) {
+            System.out.println("☹ OOPS!!! I'm sorry but I don't know what that means :-(");
+            ui.printLine();
```
Should have indentations here.

LapisRaider commented on src/main/java/Cleese/Parser.java:
```diff
@@ -0,0 +1,89 @@
+        String[] input = string.split(" ", 2);
+        String command = input[0];
+        String parameters = null;
+        if (input.length > 1) {
+            parameters = input[1];
+        }
+        switch (command) {
+        case "bye":
```
Might want to have a linebreak in between the if statement closing bracket and the switch statement to improve readability.

likeabowx replied:
I agree with @LapisRaider, in general it is easier to read code with break lines between different sections

LapisRaider commented on src/main/java/Cleese/Parser.java:
```diff
@@ -0,0 +1,89 @@
+                    newTask = new Deadline(split[0], localDateTime);
+                }
+            }
+            taskList.add(newTask);
```
Might want to put some linebreaks here to improve readability.

LapisRaider commented on src/main/java/Cleese/TaskList.java:
```diff
@@ -0,0 +1,35 @@
+    public boolean removeFromTaskList(Task selectedTaskToDelete) {
+        return taskList.remove(selectedTaskToDelete);
+    }
+    public Task get(Integer i) {
```
Might want to rename the function instead to be more descriptive.
```suggestion
    public Task getTask(Integer i) {
```

LapisRaider commented on src/main/java/Cleese/TaskList.java:
```diff
@@ -0,0 +1,35 @@
+    public Task get(Integer i) {
+        return taskList.get(i);
+    }
+    public void add(Task task) {
```
Might want to rename the function to be more descriptive.

```suggestion
    public void addTask(Task task) {
```

LapisRaider commented on src/main/java/Cleese/TaskList.java:
```diff
@@ -0,0 +1,35 @@
+    public void add(Task task) {
+        taskList.add(task);
+    }
+    public Integer size() {
```
Might want to rename the function to be more descriptive.
```suggestion
    public Integer getSize() {
```

LapisRaider commented on src/main/java/Cleese/TaskList.java:
```diff
@@ -0,0 +1,35 @@
+public class TaskList {
+    private final ArrayList<Task> taskList;
+
+    public TaskList() {
+        this.taskList = new ArrayList<>();
+    }
+    public void printTaskList() {
+        for (int i = 0; i < taskList.size(); i++) {
+            System.out.print(String.format("%d. %s \n", i+1, taskList.get(i).toString()));
+        }
+    }
+    public boolean removeFromTaskList(Task selectedTaskToDelete) {
+        return taskList.remove(selectedTaskToDelete);
+    }
+    public Task get(Integer i) {
+        return taskList.get(i);
+    }
+    public void add(Task task) {
+        taskList.add(task);
+    }
+    public Integer size() {
+        return taskList.size();
+    }
+    public void find(String keyword) {
+        for (int i = 0; i < taskList.size(); i++) {
+            if (taskList.get(i).toString().contains(keyword)) {
+                System.out.print(String.format("%d. %s \n", i+1, taskList.get(i).toString()));
+            }
+        }
+    }
```
Might want to have linebreaks in between functions to improve readability.

LapisRaider commented on src/main/java/Cleese/Task.java:
```diff
@@ -0,0 +1,32 @@
+package Cleese;
+
+public class Task {
+    protected String description;
+    protected boolean isDone;
```
Might want to change this to private instead of protected. For java, unlike C++ where protected makes the variable inaccessible to other classes besides the derived classes, in java, as long as a class is in the same package, they can access the protected variable.

If that is not your intention, you may want to change to private.

LapisRaider commented on src/main/java/Cleese/Storage.java:
```diff
@@ -0,0 +1,50 @@
+    private static String filePath;
+    private static File file;
+
+    public Storage(String filePath) {
+        this.filePath = filePath;
+        this.file = new File(filePath);
+    }
```
The usage of `this.` keyword. It is recommended to use the `this` keyword only when a field is shadowed by a method or constructor parameter.

![image](https://user-images.githubusercontent.com/36361212/152059587-025c57ef-25c7-4b34-8f0e-4e1c87d32fb0.png)

LapisRaider reviewed (commented):
Looks great! Just a few nits to fix.

TLDR:
- `this` keyword is recommended to be used only when a field is shadowed by a method or constructor parameter.
- Protected might want to change to private. Protected keyword in Java makes the variable/function accessible for any class in the same package.
- Missing some indentations.
- Might want to add more linebreaks to improve readability.
- Some suggestions on function names to make them more descriptive.

likeabowx commented on src/main/java/Cleese/Parser.java:
```diff
@@ -0,0 +1,89 @@
+        case "delete":
+            int deleteSelectedTaskNum = Integer.parseInt(parameters);
+            System.out.println(deleteSelectedTaskNum);
+            Task deleteSelectedTask = taskList.get(deleteSelectedTaskNum-1);
```
Variable names should preferably be a noun for clarity, for example taskSelectedToDelete. In this case, deleteSelectedTask sounds like a verb

likeabowx commented on src/main/java/Cleese/Task.java:
```diff
@@ -0,0 +1,32 @@
+package Cleese;
+
+public class Task {
```
Since this class is never used as an object itself, can consider making this an abstract class instead and have the other classes inherit from it. In that case, the ```toStorageString()``` method can be an abstract method and there will be no need for a ```"placeholder"``` in it.

likeabowx commented on src/main/java/Cleese/TaskList.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.ArrayList;
+
+public class TaskList {
+    private final ArrayList<Task> taskList;
```
The name of the field can be called tasks instead as it can be confused with the class name and be prone to mistakes referring to the class and object.

likeabowx reviewed (commented):
LGTM. No major problem, just some minor suggestions for improvement :slightly_smiling_face:

SethCKL committed:
add javadocs and fix codestyle

SethCKL committed:
update codestyle

SethCKL committed:
remove debug statement from parser

SethCKL committed:
no message

SethCKL committed:
Merge branch 'add-gradle-support'

SethCKL committed:
added level 10

SethCKL committed:
change gradle build file and add checkstyle xml

SethCKL committed:
Add assert statements to Cleese and Ui classes

The Cleese and Ui class are crucial to get the program to a
state where we can test basic functionality and correctness.

Adding assert statements will help to ensure that the classes
and underlying functions are always given the correct input.

This allows for increased confidence of bugs or issues
arising from other parts of the program and speeds up
potential debugging.

SethCKL committed:
improve code quality across all files where applicable

SethCKL committed:
add help command to assist user in knowing what commands are available

SethCKL committed:
add javadocs to methods flagged by checkstyle

SethCKL committed:
Merge pull request #2 from SethCKL/branch-A-CodeQuality

add javadocs to methods flagged by checkstyle

SethCKL committed:
add Ui.png

SethCKL committed:
Update README.md

SethCKL committed:
Update README.md

SethCKL committed:
Update README.md

SethCKL committed:
Update README.md

SethCKL committed:
Update README.md

SethCKL committed:
update build.gradle CleeseTest.java

SethCKL committed:
Merge branch 'master' of https://github.com/SethCKL/ip

SethCKL committed:
update to fix storage issue

SethCKL committed:
no message

SethCKL committed:
Update UserGuide

SethCKL committed:
Rename CleeseTest Package and edit launcher javadocs

SethCKL committed:
Improve Cleese

Edit documentation and user guide
Increase abstraction
Tweak Ui
Adjust Unit Test
